flocker/control/test/test_model_change.py

```diff
@@ -0,0 +1,418 @@
+"""
+
+from json import loads, dumps
+from subprocess import check_output
+
+from pyrsistent import (
+    PRecord, PClass, CheckedPSet, CheckedPVector, CheckedPMap, field,
+)
+
+from twisted.trial.unittest import SynchronousTestCase
+from twisted.python.filepath import FilePath
+from twisted.python.reflect import qual as fqpn
+
+from .._model import Deployment
+
+PERSISTED_MODEL = FilePath(__file__).sibling(b"persisted_model.json")
+
+# The class at the root of the configuration tree. This may need to be
+# changed if the configuration's root class changes.
+ROOT_CLASS = Deployment
+
+
+def generate_model(root_class=ROOT_CLASS):
+    """
+    Generate a data-structure that represents the current configuration
+    model.
+
+    Changes to output may require regenerating the persisted version
+    on-disk.
+
+    This should switch to Pyrsistent's introspection API once it exists:
+    https://github.com/tobgu/pyrsistent/issues/47
+    """
+    classes_result = {}
+    result = {u"root": fqpn(root_class),
+              u"classes": classes_result}
+    classes = {root_class}
+    while classes:
+        klass = classes.pop()
+        klass_name = fqpn(klass)
+        if klass_name in classes_result:
+            continue
+        record = None
+        if issubclass(klass, (PRecord, PClass)):
+            if issubclass(klass, PRecord):
+                attr_name = "_precord_fields"
+            else:
+                attr_name = "_pclass_fields"
+            record = {u"category": u"record",
+                      u"fields": {}}
+            for name, field_info in getattr(klass, attr_name).items():
+                record[u"fields"][name] = list(
+                    fqpn(cls) for cls in field_info.type)
+                for cls in field_info.type:
+                    classes.add(cls)
+        elif issubclass(klass, CheckedPMap):
+            record = {
+                u"category": u"map",
+                u"fields": {
+                    u"key": list(
+                        fqpn(cls) for cls in klass._checked_key_types),
+                    u"value": list(
+                        fqpn(cls) for cls in klass._checked_value_types)}}
+            for cls in klass._checked_key_types + klass._checked_value_types:
+                classes.add(cls)
+        elif issubclass(klass, (CheckedPSet, CheckedPVector)):
+            category = u"set" if issubclass(klass, CheckedPSet) else u"list"
+            record = {
+                u"category": category,
+                u"type": list(fqpn(cls) for cls in klass._checked_types),
+            }
+            for cls in klass._checked_types:
+                classes.add(cls)
+        classes_result[klass_name] = record
+    return result
+
+
+class Subtype(PRecord):
+    """
+    A sub-type used in ``GenerateModelTests``.
+    """
+
+
+class ChangedSubtype(PRecord):
+    """
+    A changed variant of ``Subtype``.
+    """
+    x = field()
+ChangedSubtype.__name__ = "Subtype"
```

Comment:
:question: I wonder if this idiom would be better:

```py
class Subtype(PRecord):
    ...
OriginalSubtype = Subtype

class Subtype(OriginalSubtype):
    ...
```

in case `__name__` isn't the on place the name gets recorded.